### ISSUE TITLE
`reason` of C++ errors are not send to Node.js

### ISSUE THREAD

**Issue by piranna** [versatica/mediasoup#748]
### Your environment

- Operating system: Ubuntu Linux 21.10
- Node version: v16.13.1
- npm version: 8.1.2
- gcc/clang version: 11.2.0-7ubuntu2
- mediasoup version: 3.9.3
- mediasoup-client version: (not applicable)

### Issue description

I'm calling to `AudioLevelObserver.removeProducer()` on Producer `close` event, so I'm getting an `Error:  [method:rtpObserver.removeProducer]` message and probably it's because the Producer is closed in C++ space and not accesible anymore, so an `Error` is being thrown at https://github.com/versatica/mediasoup/blob/34dcdcd157f8d6966d293e5b61272d1845688711/worker/src/RTC/Router.cpp#L528-L529, but not fully sure yet. Problem is, the `Producer not found` error message is not being printed anywhere (note the two spaces at `:  [` in the provided error message).

I've been investigating the C++ side source code trying to find the source of this problem, not sure if my conclussions are correct, please correct me if I'm wrong.

So, that error is thrown by the [MS_THROW_ERROR](https://github.com/versatica/mediasoup/blob/1e3bdac019f4507652581480cb6f9cdbdebbdb5f/worker/include/MediaSoupErrors.hpp#L29-L35) macro, that's using `std::snprintf()` function to write the `Producer not found` error description in a `thread_local` Exception variable, common to all `MediaSoupError` instances, that's used to define the exception `reason`.

This exception is being catch at https://github.com/versatica/mediasoup/blob/f0ebd8099a7bad12d55f9ff017a9a6b175e1820d/worker/src/Worker.cpp#L313-L316, that's throwing a new exception itself calling again to `MS_THROW_ERROR` macro to compose the error reason with the method name to know who was the one that throwed it, and later is finally catch and serialized as an error to the Node.js side at https://github.com/versatica/mediasoup/blob/f0ebd8099a7bad12d55f9ff017a9a6b175e1820d/worker/src/Channel/ChannelSocket.cpp#L251-L254.

Now, problem is, since `Worker::OnChannelRequest()` function is throwing an exception fetching the reason from the original one initially thrown by `Router::GetProducerFromData()`, depending of the implementation of `std::runtime_error`, if it's not being copied (please correct me if I'm wrong) then the variable `MediaSoupError::buffer` is being used both as input and target variable of `std::snprintf()`, and according to https://en.cppreference.com/w/cpp/io/c/fprintf:

> If a call to sprintf or snprintf causes copying to take place between objects that overlap, the behavior is undefined (e.g. sprintf(buf, "%s text", buf);)

So, being undefined behaviour, it makes sense that it interpret the content as null string, for example if `std::snprintf()` zeroed the target variable. So, if that's the case, then probably we would need to allocate an intermediary variable and use it instead to do the exception reason composition. Does it makes sense, or I'm overlooking something? Maybe the error of not showing the original exception reason is due to anything else?

**Comment by ibc** [comment]
> Problem is, the Producer not found error message is not being printed anywhere (note the two spaces at : [ in the provided error message).

Which provided error message? I don't see it in the issue description.

**Comment by piranna** [comment]
> Which provided error message? I don't see it in the issue description.

The `Error: [method:rtpObserver.removeProducer]` string at beginning. Full error trace is

```js
Error:  [method:rtpObserver.removeProducer]
     at Channel.processMessage (/app/node_modules/mediasoup/node/lib/Channel.js:195:37)
     at Socket.<anonymous> (/app/node_modules/mediasoup/node/lib/Channel.js:69:34)
     at Socket.emit (node:events:390:28)
     at addChunk (node:internal/streams/readable:315:12)
     at readableAddChunk (node:internal/streams/readable:289:9)
     at Socket.Readable.push (node:internal/streams/readable:228:10)
     at Pipe.onStreamRead (node:internal/stream_base_commons:199:23)
``` 

Sorry for not put it in first place, didn't consider it important, but you are right, it's needed to know exactly where it's being thrown.

**Comment by piranna** [comment]
Reviewing the Mediasoup `debug` logs, I've found a more detailed case where this is also happening:

```
2022-01-07T07:07:15.046Z mediasoup:ERROR:Channel [pid:45 RTC::Router::GetProducerFromData() \| throwing MediaSoupError: Producer not found
2022-01-07T07:07:15.046Z mediasoup:ERROR:Channel [pid:45 Worker::OnChannelRequest() \| throwing MediaSoupError: Producer not found [method:rtpObserver.removeProducer]
2022-01-07T07:07:15.046Z mediasoup:WARN:Channel request failed [method:rtpObserver.removeProducer, id:1747]:  [method:rtpObserver.removeProducer]
```

This is the actual error messages I'm getting by calling `remoteProducer()` after clossing the Producer instance: first two ones are ok, with the `Producer not found` string, but the third one that's replaced by an empty string, look at the two spaces at the middle of `id:1747]:  [method`, the error string should be placed between them but it's like it's interpret as an empty string instead.

**Comment by piranna** [comment]
(after that messages, it's printed the stack trace of previous comment)

**Comment by ibc** [comment]
Fixed in PR #750

**Comment by piranna** [comment]
Thank you :-)
